File: backend/.NET/Calindra.Desafio/Calindra.Desafio.Domain/Results/Result.cs
namespace Calindra.Desafio.Domain.Results
{
    public static class Result
    {
        public static MethodResult Ok()
            => new();

        public static MethodResult Ok(object data)
            => new(data);

        public static MethodResult Fail(string propertyName, params string[] messages)
            => new(propertyName, messages);

        public static MethodResult FailWithData(string propertyName, object data, params string[] messages)
        {
            return new MethodResult(propertyName, messages)
            {
                Data = data
            };
        }

        public static MethodResult Parse<T>(MethodResult<T> methodResult)
        {
            var result = new MethodResult(methodResult.Data);
            if (methodResult.Failure)
                result.Add(methodResult.Failures);
            return result;
        }
    }
}

File: backend/Node.JS/calindra-desafio-serverless/src/shared/results/result.ts
import { MethodResult } from './method-result';

export class Result {
    
    public static ok(data: object | null = null) {
        return Result.default(data);
    } 

    public static fail(propertyName: string, ...messages: string[]) {
        const methodResult = new MethodResult();
        methodResult.add(propertyName, ...messages);
        return methodResult;
    }

    public static failWithData(data: object, propertyName: string, ...messages: string[]) {
        const methodResult = Result.default(data);
        methodResult.add(propertyName, ...messages);
        return methodResult;
    }

    private static default(data: object | null = null) {
        const methodResult = new MethodResult();
        if (data != null)
            methodResult.data = data;
        return methodResult;
    }
}